src/Perlang.Interpreter/PerlangInterpreter.cs

```diff
@@ -569,30 +569,43 @@ private void Interpret(IEnumerable<Stmt> statements)
             return expr.Value;
         }
 
-        public object? VisitLogicalExpr(Expr.Logical expr)
+        public object VisitLogicalExpr(Expr.Logical expr)
         {
-            object? left = Evaluate(expr.Left);
+            // Note: ! operators in this method are considered "safe" since the type validation is expected to have
+            // validated that these operators are only called with `bool` operands.
```

Comment:
I'm not 100% sure this is a good idea or not... It essentially makes some parts of the code blindly trust other parts of the code. This _can_ be a horrible idea, or perhaps it's totally genius. Time will tell. :slightly_smiling_face: